### ISSUE TITLE
Investigate server-side apply to allow cooperating controllers

### ISSUE THREAD

**Issue by dprotaso** [knative/pkg#2128]
Refs: 
https://kubernetes.io/docs/reference/using-api/server-side-apply/
https://github.com/knative-sandbox/net-contour/issues/347#issuecomment-838613107

This would allow customization of the resources we create

/assign @dprotaso

**Comment by dprotaso** [comment]
FYI there's a server-side apply bug upstream that's worth keeping tabs on - https://github.com/kubernetes/kubernetes/issues/111937

I don't believe if blocks this issue

**Comment by enarha** [comment]
I want to confirm this is still something of interest. I created a PR  https://github.com/tektoncd/pipeline/pull/9013 to add Server-Side Apply patches for managing finalizers in Tekton Pipeline and it was suggested I add that functionality here instead. Would such contribution be welcome?

**Comment by dprotaso** [comment]
Hey @enarha - last I looked at SSA it didn't support unit testing using fakes clients etc. Has that changed?


Looking at the PR it seems straightforward addition to the generator. Though I'm curious when would get a conflict when you delete a finalizer?

**Comment by enarha** [comment]
I tested that with two controllers (Tekton Chains and Tekton Results) operating on their own finalizer. Given each controller uses a different fieldManager string, there was no issues with removing the finalizer for neither of them. Quite the opposite when comparing with the current merge patch + resourceVersion combination, where I see 3-4 retries on each controller until they manage to set/remove a finalizer.
One edge case I found compared to merge patch is the different behavior when we operate on stale object from the informer's cache. Using merge patch that's permanent failure (NotFound), SSA though tries to create the object when it's missing in etcd, leading to admission or RBAC failures, which is a not permanent error followed by more retries until the object is removed from the cache. I had documented that in my PR and implemented optimization for it.
I did not have a chance to look into the fake clients, but I'll do and update.

**Comment by enarha** [comment]
> Hey [@enarha](https://github.com/enarha) - last I looked at SSA it didn't support unit testing using fakes clients etc. Has that changed?
> 
Yes, that has been improved. Apply support has been added to fakeclient https://github.com/kubernetes-sigs/controller-runtime/pull/2981 and client-go fake cleint also supports field manager.

> Looking at the PR it seems straightforward addition to the generator. Though I'm curious when would get a conflict when you delete a finalizer?

In the general case where the finalizer is set using SSA, there is no problem to remove it. The same (controller) field manager who set the finalizer sends a request omitting that finalizer which leads to removal of that finalizer. Any existing finalizers not managed by the same field manager won't be affected. There is an edge case though with objects in flight caught during the switch from merge patch strategy to SSA. The SSA can't remove the finalizer on those objects because it does not own them (even if called with --force-conflicts=true). There is no actual conflict (409), it's just a no-op and the finalizer is not removed. Those would have to be manually removed using merge patch.
